### PR TITLE
Support makie backends

### LINKED ISSUE #54: Allow using other Makie backends
Right now, only GLMakie backend is allowed, for e.g. WGLMakie or CairoMakie, RecipesMakie.jl is not loaded (as per [this line](https://github.com/rveltz/BifurcationKit.jl/blob/4bc09399d790f7b23c0fcfeb38808db968fb5ea3/src/BifurcationKit.jl#L84))

Also, RecipesMakie.jl seems to only support GLMakie (but that's only my assumption).

Thanks a lot

### PR DESCRIPTION
**Pull request by BambOoxX**
This is a first proposal to allow other Makie backends to be used with BifurcationKit. 
The first main problem (solved in this draft) is that (un)stable branches are displayed with different line widths, which is incompatible with CairoMakie as is. 
Another solution could be to have one plot per branch. But I'm not sure what is best.

Fixes #54 for CairoMakie at least.

### CODE REVIEW

File: ext/MakieExt/plot.jl [hunk 1]
```diff
@@ -1,147 +1,172 @@
-using GLMakie: Point2f0
+using Makie: Point2f0
 
-function GLMakie.convert_arguments(::PointBased, contres::AbstractBranchResult, vars = nothing, applytoY = identity, applytoX = identity)
+function Makie.convert_arguments(::PointBased, contres::AbstractBranchResult, vars=nothing, applytoY=identity, applytoX=identity)
```
Comment (rveltz) [inline ext/MakieExt/plot.jl:3]: My coding style is to have spaces around =

File: ext/MakieExt/plot.jl [hunk 2]
```diff
@@ -1,147 +1,172 @@
...
 function plot!(ax1, contres::AbstractBranchResult;
-                plotfold = false,
-                plotstability = true,
-                plotspecialpoints = true,
-                putspecialptlegend = true,
-                filterspecialpoints = false,
-                vars = nothing,
-                linewidthunstable = 1.0,
-                linewidthstable = 3.0linewidthunstable,
-                plotcirclesbif = true,
```
Comment (rveltz) [inline ext/MakieExt/plot.jl:17]: My coding style is to have spaces around =

**Comment by rveltz** [comment]
Wow, thank you for this new PR!!

One issue I found is that the following does not work

```
function TMvf!(dz, z, p, t = 0)
    (;J, α, E0, τ, τD, τF, U0) = p
    E, x, u = z
    SS0 = J * u * x * E + E0
    SS1 = α * log(1 + exp(SS0 / α))
    dz[1] = (-E + SS1) / τ
    dz[2] = (1.0 - x) / τD - u * x * E
    dz[3] = (U0 - u) / τF +  U0 * (1 - u) * E
    dz
end

par_tm = (α = 1.5, τ = 0.013, J = 3.07, E0 = -2.0, τD = 0.200, U0 = 0.3, τF = 1.5, τS = 0.007)
z0 = [0.238616, 0.982747, 0.367876 ]
prob = BifurcationProblem(TMvf!, z0, par_tm, (@optic _.E0); record_from_solution = (x, p; k...) -> (E = x[1], x = x[2], u = x[3]),)

opts_br = ContinuationPar(p_min = -10.0, p_max = 1., dsmax = 0.1, n_inversion = 8, nev = 3)
br = continuation(prob, PALC(tangent = Bordered()), opts_br; plot = false, normC = norminf, bothside = true)

BK.plot(br, plotfold=false)
####################################################################################################
br_fold = BK.continuation(br, 2, (@optic _.α), ContinuationPar(br.contparams, p_min = 0.2, p_max = 5.), bothside = true)
BK.plot(br_fold)
```

File: ext/MakieExt/plot.jl (hunk 1 above)
Comment (BambOoxX) [reply ext/MakieExt/plot.jl:3]: Ah yes, sorry about that, I hit auto indent and forgot to check it. I just don't know yet in vscode how to change the formatting settings. Would you know ?

Comment (rveltz) [reply ext/MakieExt/plot.jl:3]: not really

**Comment by BambOoxX** [comment]
> br_fold = BK.continuation(br, 2, (@optic _.α), ContinuationPar(br.contparams, p_min = 0.2, p_max = 5.), bothside = true)
> BK.plot(br_fold)

This works on latest.

**Comment by rveltz** [comment]
Yep but this one is broken now :(

```julia
# test vector field for event detection
function Feve(X, p)
	p1, p2, k = p
	x, y = X
	out = similar(X)
	out[1] = p1 + x - y - x^k/k
	out[2] = p1 + y + x - 2y^k/k
	out
end

# parameters for the vector field
par = (p1 = -3., p2 = -3., k = 3)

# bifurcation problem
prob = BifurcationProblem(Feve, -2ones(2), par, (@optic _.p1);
	record_from_solution = (x, p; k...) -> x[1])

# parameters for the continuation
opts = ContinuationPar(max_steps = 150, p_min = -3., p_max = 4.0,
    # we disable bifurcation detection to focus
    # on the user passed events
    detect_bifurcation = 0, detect_fold = false,
    # parameters specific to event detection
    detect_event = 2)
# arguments for continuation
args = (prob, PALC(), opts)
kwargs = (plot = false, verbosity = 0,)
br = continuation(args...; kwargs...,)
BK.plot(br)


br = continuation(args...; kwargs...,
	event = SaveAtEvent((-2.,-1.); use_newton = true))

BK.plot(br)
```

**Comment by BambOoxX** [comment]
It should be okay now, it was because empty `indices` were not handled properly. Now any edge case should be okay. Although I can't test everything easily, sorry.

**Comment by rveltz** [comment]
Do you think it is good for merging?
I have tried deflated continuation, diagram and codim2, it seems to work well with CairoMakie

**Comment by BambOoxX** [comment]
At this point, I think it works. The only question is if there are other uses of lifestyles I did not meet during my tests. It's really the only problem I met with CairoMakie.
Also, I was thinking about a function capable of altering the backend `set_plot_backend!(::AbstracPlotBackend)` so that people could switch between Makie and Plots upon request.
Do you want me to add a commit with that before merging ?

**Comment by rveltz** [comment]
yes if you have a working solution, that would be awesome.
Thank you a lot for your work, it's really improved BK.

The Makie backend is slower than the Plots one during continuation. I think the Makie backend should rely on observables in that case but that's another issue.

**Comment by BambOoxX** [comment]
> yes if you have a working solution, that would be awesome.

So actually, it doesn't exactly require `set_plot_backend!`, `BK.plot` will plot with Makie backends and `Plots.plot` with Plots backend, although it would be better to have `BK.plot` work for all backends.
> Thank you a lot for your work, it's really improved BK.

Thanks, but compared to all the difficult math stuff, this is very easy ;).
> The Makie backend is slower than the Plots one during continuation. I think the Makie backend should rely on observables in that case but that's another issue.

Yep I could work on that if you have examples that show the problem.

**Comment by rveltz** [comment]
> Yep I could work on that if you have examples that show the problem.

Not really, when trying both backend on a simple example, I came to this subjective observation

**Comment by rveltz** [comment]
You might also be interested by https://github.com/gszep/BifurcationExplorer.jl
